### ISSUE TITLE
Large App Size

### ISSUE THREAD

**Issue by ceremonious** [bennymeg/nx-electron#75]
Hi, I packaged my React frontend into an Electron app and the output size is around ~3 GB. I'm not running any backend NodeJS processes in Electron, other than a light script to handle sending notifications via Firebase. The app.asar file is 2.78 GB.

I've never used Electron before, but from what I've seen it seems like Electron includes all of node_modules in the final package, instead of just the modules required by the React frontend. Because I'm using NX as a monorepo for multiple projects, including a React Native app & a NodeJS backend, my node_modules is around 2.45 GB. Is there a reason that Electron needs to include all of node_modules? If I exclude node_modules in maker.options.json (via the files option), will I end up running into issues? Is there another config setting somewhere I can change to reduce the final bundle size?

Thanks!

**Comment by bennymeg** [comment]
Hi,
- Nx uses only single `package.json` file, so all the dependencies in that file would be installed when you package you app.
  This is a know disadvantage of Nx ([#1777](https://github.com/nrwl/nx/issues/1777)). You can try using yarn workspaces or similar work arounds the the Nx community 
   suggesting.
- When you package your app, `nx-electron` should ignore all your development and non essential electron dependencies.
- Your issue is probably related to angular, it use native gRPC package and when in recompiled against electron-abi it grows to 
   more then 1GB. try reinstalling angular and using this configurations at your `nx-electron` maker config file:
   ```
  "electronCompile": false,
  "npmRebuild": false,
   ```

**Comment by 3Shain** [comment]
> Hi,
> 
> * Nx uses only single `package.json` file, so all the dependencies in that file would be installed when you package you app.
>   This is a know disadvantage of Nx ([#1777](https://github.com/nrwl/nx/issues/1777)). You can try using yarn workspaces or similar work arounds the the Nx community
>   suggesting.
> * When you package your app, `nx-electron` should ignore all your development and non essential electron dependencies.
> * Your issue is probably related to angular, it use native gRPC package and when in recompiled against electron-abi it grows to
>   more then 1GB. try reinstalling angular and using this configurations at your `nx-electron` maker config file:
>   ```
>   "electronCompile": false,
>   "npmRebuild": false,
>   ```

Could you check this out [https://nx.dev/latest/angular/plugins/node/executors/build#generatepackagejson](url) , I think it could be implemented for nx-electron as well. It would be nice to have a standalone package.json in app folder.

**Comment by 3Shain** [comment]
I think I've got a work around:

add this in your directory:
```json
"directories":{
    "app": "dist/apps/<YOUR APP>"
  }
```

and then move the **package.json with only necessary dependencies** to this folder (and global package-lock.json if you want ), then `nx <YOUR APP>:<make/package> [--options]`. In my case the bundle size is reduced by 40%.

However, be sure every time you make or package your application you should remove your old build folder and do this approach from `nx build <YOUR APP>`, otherwise any change in dependencies might not be applied. Make this as a CI instruction is better IMO.

**Comment by 3Shain** [comment]
> I think I've got a work around:
> 
> add this in your maker.options.json:
> 
> ```json
> "directories":{
>     "app": "dist/apps/<YOUR APP>"
>   }
> ```
> 
> and then move the **package.json with only necessary dependencies** to this folder (and global package-lock.json if you want ), then `nx <YOUR APP>:<make/package> [--options]`. In my case the bundle size is reduced by 40%.
> 
> However, be sure every time you make or package your application you should remove your old build folder and do this approach from `nx build <YOUR APP>`, otherwise any change in dependencies might not be applied. Make this as a CI instruction is better IMO.

So it would be great if a `generatePackageJson` like feature could be implemented , although it's ok to write package.json manually.

**Comment by bennymeg** [comment]
@3Shain it seems that [Nx implementation](https://github.com/nrwl/nx/blob/master/packages/node/src/utils/generate-package-json.ts) takes care only on the package it was used in. nx-electron takes care for the frontend projected as well, I will see if that could be implemented in the future.

I am not sure that changing the app directory is a good solution (it might be used for other things then packageJson directory - but I might be wrong, I need to investigate it a bit more). If that works for you, you can always extend the webpack config in similar way to [this implementation](https://github.com/nrwl/nx/issues/1518#issuecomment-533555445).

**Comment by 3Shain** [comment]
@bennymeg Normally frontend project should be bundled (as there is no way to access local 'node_modules' files in browser) so actually we should only take care of the 'backend' part. Theoretically backend could be also bundled to a single js file, however this method doesn't support native addons and dynamic package import.

**Comment by ceremonious** [comment]
So I had initially misconfigured the files parameter in maker.options.json, which was causing unnecessary files to be bundled. After fixing that, my application size went down to ~800 MB, which was still much higher than I would like due to including node_modules. 

I've taken a similar approach to 3Shain and am using a secondary package.json that only includes the node_modules required by the Electron app. Right now, this package.json needs to be manually curated, which is not ideal for future updates. But the # of dependencies my Electron app has is low, so it's manageable for now.

**Comment by bennymeg** [comment]
@3Shain I agree, but for some reason the was not always worked that way with electron.
@ceremonious that's great, did you got to check the grpc module size?

**Comment by ceremonious** [comment]
I didn't check the grpc module size because I'm using React, not angular

**Comment by bennymeg** [comment]
I am not sure way firebase do not use gRPC in react, never mind.

Anyway, I have added `generatePackageJson` option to the build schematic, let me know how it works for you.

**Comment by julianpoemp** [comment]
I'm facing the same issue with the latest version of nx-electron, nx and angular. I created a new project with Angular for frontend. I set the following maker options:

**maker.options.json**
````json
{
  "$schema": "../../../../../node_modules/nx-electron/src/validation/maker.schema.json",
  "directories": {
    "app": "dist/apps/electron-app"
  },
  "files": [
    "index.html"
  ]
}
````

My angular app is called `nx-electron-test` so I did the following steps:

1. nx build nx-electron-test && nx build electron-app
2. Create a file `package.json` with as small information as possible to `./dist/apps/electron-app`. Somehow "generatePackageJson" seems not working:

````json
{
  "name": "electron-app",
  "version": "1.0.0",
  "dependencies": {},
  "devDependencies": {}
}
````

3. nx run electron-app:package --platform=mac --arch=x64 --prepackageOnly=false

Result: macOS app file of size 238MB. Before setting the options in maker.options.json and creating the package.json manually it was ~438 MB.

The app size is too big for fresh initialized project without adding any third party dependencies. If ElectronJS consumes ~100MB the file size should be around that size and not 238MB.

Are there any new methods to keep the app size as small as possible?
